purchase_order_price_recalculation/models/purchase_order.py

```diff
@@ -0,0 +1,26 @@
+# Copyright 2022 Tecnativa - Víctor Martínez
+# License AGPL-3 - See http://www.gnu.org/licenses/agpl-3.0.html
+from odoo import api, models
+
+
+class PurchaseOrder(models.Model):
+    _inherit = "purchase.order"
+
+    @api.model
+    def _recompute_form_line_vals(self, fields):
+        """Helper method for writing in line certain recomputed by onchanges fields."""
+        for line in self.mapped("order_line"):
+            vals = line._convert_to_write(line.read()[0])
+            if "product_tmpl_id" in line._fields:
+                vals["product_tmpl_id"] = line.product_tmpl_id
+            # we make this to isolate changed values:
+            line2 = self.env["purchase.order.line"].new(vals)
+            line2.env.add_to_compute(
+                line2.product_id._fields["price"], line2.product_id
+            )
+            line2.onchange_product_id()
+            line.write({field: line2[field] for field in fields})
```

Comment:
I'd say a better way to recalculate the lines would be to just trigger `_onchange_quantity` line by line. This way, you'll get the whole pack: discounts, new reception days and anything that could come from the supplierinfo.

https://github.com/odoo/odoo/blob/421e9d9100e9593999c9490342070a81b1786078/addons/purchase/models/purchase.py#L1071